x11rb-async/src/rust_connection/mod.rs

```diff
@@ -0,0 +1,488 @@
+);
+
+/// A connection implemented entirely in pure Rust.
+pub struct RustConnection<S: Stream = DefaultStream> {
+    /// The inner connection.
+    inner: Arc<InnerConnection<WrappedStream<S>>>,
+
+    /// Prevent more than one task from reading at once.
+    read_lock: Mutex<()>,
+
+    /// Prevent more than one task from writing at once.
+    write_lock: Mutex<()>,
+}
+
+/// Wraps a stream registered in the reactor.
+///
+/// `CURRENT_CONTEXT` should be set when calling this; otherwise a panic may occur.
+struct WrappedStream<S: Stream> {
+    /// Register the stream into the reactor.
+    stream: Async<S>,
+}
+
+impl<S: Stream> Stream for WrappedStream<S> {
+    fn poll(&self, mode: PollMode) -> io::Result<()> {
+        // Get the current context.
+        CURRENT_CONTEXT.with(|waker| {
+            let mut ctx = Context::from_waker(waker);
+
+            // Poll the stream.
+            let poll = match mode {
+                PollMode::Readable => self.stream.poll_readable(&mut ctx),
+                PollMode::Writable => self.stream.poll_writable(&mut ctx),
+                PollMode::ReadAndWritable => {
+                    // Poll both at once.
+                    let readable = self.stream.poll_readable(&mut ctx);
+                    let writable = self.stream.poll_writable(&mut ctx);
+
+                    // Combine the polls.
+                    match (readable, writable) {
+                        (Poll::Ready(Ok(())), Poll::Ready(Ok(()))) => Poll::Ready(Ok(())),
+                        (Poll::Ready(Err(e)), _) | (_, Poll::Ready(Err(e))) => Poll::Ready(Err(e)),
+                        _ => Poll::Pending,
+                    }
+                }
+            };
+
+            // Convert the result.
+            match poll {
+                Poll::Ready(res) => res,
+                Poll::Pending => Err(io::ErrorKind::WouldBlock.into()),
+            }
+        })
+    }
+
+    // Forward remaining methods down lower, it's non-blocking anyways.
+
+    fn read(&self, buf: &mut [u8], fd_storage: &mut Vec<RawFdContainer>) -> io::Result<usize> {
+        self.stream.get_ref().read(buf, fd_storage)
+    }
+
+    fn read_exact(&self, buf: &mut [u8], fd_storage: &mut Vec<RawFdContainer>) -> io::Result<()> {
+        self.stream.get_ref().read_exact(buf, fd_storage)
+    }
+
+    fn write(&self, buf: &[u8], fds: &mut Vec<RawFdContainer>) -> io::Result<usize> {
+        self.stream.get_ref().write(buf, fds)
+    }
+
+    fn write_vectored(
+        &self,
+        bufs: &[io::IoSlice<'_>],
+        fds: &mut Vec<RawFdContainer>,
+    ) -> io::Result<usize> {
+        self.stream.get_ref().write_vectored(bufs, fds)
+    }
+}
+
+impl RustConnection {
+    /// Connect to the X11 server.
+    pub async fn connect(dpy_name: Option<&str>) -> Result<(Self, usize), ConnectError> {
+        let (stream, screen) = nb_connect::connect(dpy_name).await?;
+        Ok((Self::connect_to_stream(stream, screen).await?, screen))
```

Comment:
No `.Xauthority` "fun"? I guess that deserves at least a `TODO` comment somewhere.

(Yay, I guess that means the Xauth code needs to be made public somewhere... do we need another crate besides `x11rb-protocol` for such shared code?